src/modules/rust.rs

```diff
@@ -119,26 +120,48 @@ fn execute_rustup_default(context: &Context) -> Option<String> {
         .map(str::to_owned)
 }
 
+// convert a path to a key in rustup `settings.toml`.
+// Mimics rustup:
+// * `settings::Settings::path_to_key`
+// * `utils::utils::canonicalize_path`
+fn path_to_key(path: &Path) -> String {
+    if path.exists() {
+        // allow disallowed method since we want to mimic
+        // rustup exactly.
+        #[allow(clippy::disallowed_methods)]
+        fs::canonicalize(path)
+            .unwrap_or_else(|_| PathBuf::from(path))
+            .display()
+            .to_string()
+    } else {
+        path.display().to_string()
+    }
+}
+
 fn extract_toolchain_from_rustup_override_list(stdout: &str, cwd: &Path) -> Option<String> {
     if stdout == "no overrides\n" {
         return None;
     }
-    // use display version of path, also allows stripping \\?\
-    let cwd = cwd.to_string_lossy();
-    // rustup strips \\?\ prefix
-    #[cfg(windows)]
-    let cwd = cwd.strip_prefix(r"\\?\").unwrap_or(&cwd);
 
-    stdout
+    let overrides: HashMap<String, &str> = stdout
         .lines()
         .filter_map(|line| {
             let (dir, toolchain) = line.split_once('\t')?;
-            Some((dir.trim(), toolchain.trim()))
+            Some((path_to_key(Path::new(dir.trim())), toolchain.trim()))
         })
-        // find most specific match
-        .filter(|(dir, _)| cwd.starts_with(dir))
-        .max_by_key(|(dir, _)| dir.len())
-        .map(|(_, toolchain)| toolchain.to_owned())
+        .collect();
+
+    let mut dir = Some(cwd);
+
+    while let Some(d) = dir {
+        let key = path_to_key(d);
```

Comment:
Does this really need `canonicalize` at every level or at all? We can be reasonably sure that the first path exists, and each call will do I/O. Both the overrides and cwd should already be canonicalized (`dunce` will just sometimes strip the prefix).

I'd prefer to just strip the prefixes and convert each level directly into a `String`.

Reply:
> I'd prefer to just strip the prefixes and convert each level directly into a `String`.

This probably boils down to a general design-decision. Personally I prefer using library features ( in this case, for paths and their comparison) over reinventing the wheel. In the end, this is what lead to this bug here since prefix-matching works different for strings and paths. 

Secondly, this mimics how `rustup` does it, so possible edge-cases I don't know about are covered. 

IMO I/O effort is negligible in comparison to the rest of starship or even the rest of the rust-module. 

**Of course if that's preferred maintainers I can fix the prefix matching while still using strings and not paths.**

> Both the overrides and cwd should already be canonicalized (dunce will just sometimes strip the prefix).

For the overrides you're probably right, since rustup probably stores the canonicalized path in `settings.yml. But when I removed the canonicalization for the overrides, the windows-test failed. How does the `settings.yml` look on windows compared to the tests?

Reply:
@davidkna could you give me some feedback here? 
I would love to fix this bug. 

Of course I'm happy to adapt to a string-based approach if that's preferred.

Reply:
>Personally I prefer using library features ( in this case, for paths and their comparison) over reinventing the wheel.

Aren't this code and rustup itself just creating string keys for each parent and comparing them? I'd prefer not doing canonicalization here, but if you feel strongly about it, you can stay with this approach.

> But when I removed the canonicalization for the overrides, the windows-test failed.

Rustup unconditionally removes the verbatim path prefix. The canonicalization adds back the prefix back to both `current_dir` and the override and makes them match (this code is doing the opposite of rustup).

Reply:
> Personally I prefer using library features ( in this case, for paths and their comparison) over reinventing the wheel.

By the way: #3414 seems to be doing proper Path-based comparison and should not be too far off.

Reply:
> > Personally I prefer using library features ( in this case, for paths and their comparison) over reinventing the wheel.
> 
> By the way: #3414 seems to be doing proper Path-based comparison and should not be too far off.

Are you sure this works on windows too? 

To me #3414 looks like it's reverting the windows-specific change you made in #3359, and just comparing a `PathBuf` override with a `Path` cwd.

Reply:
> Rustup unconditionally removes the verbatim path prefix. The canonicalization adds back the prefix back to both `current_dir` and the override and makes them match (this code is doing the opposite of rustup).

I thought I read the rustup code correctly and mimicked its behaviour. I will check this again.

Reply:
> Are you sure this works on windows too?

There's room for improvement, but it should work most of the time because the verbatim path prefix is usually stripped by `dunce` now.

Reply:
@davidkna ah, I think I understand now, please correct me if I'm wrong (and sorry for the long discussion): 

The introduction of `dunce` ( #3638 ) actually makes stripping the prefix in #3359 unnecessary because the prefix is stripped in the `Context`. Correct? 

Then I can just change this PR to use normal `Path::starts_with` again.

Reply:
Stripping the prefix is still somewhat more correct, because paths that are potentially invalid without the verbatim prefix will not have it stripped by dunce, whereas IIRC rustup is doing it unconditionally.

It would be better to convert the Path to `&str`, strip the prefix and convert it back to `Path` for `Path::starts_with`.